### PR TITLE
Adds Violations Requirements to MoneyRequestConfirmationList

### PR DESCRIPTION
**Pull request by cdanwards**
### Fixed Issues

$ https://github.com/Expensify/App/issues/31090
PROPOSAL: 

#### For Categories:
- Go to `Settings > Workspaces > [the workspace] > Categories`.
- Turn on "People must categorize expenses".
- Ensure at least one category is enabled.

#### For Tags:
- Go to `Settings > Workspaces > [the workspace] > Tags`.
- Turn on "People must tag expenses".
- Ensure at least one tag is enabled. Make sure "Do you want to use multiple levels of" is off.

#### General Steps:
- [ ] Verify that no errors appear in the JS console.
- [ ] Set the `canUseViolations` Permissions to true within the `Permissions` file.

#### Case 1: Neither Categories nor Tags Required
- [ ] Follow steps to disable both categories and tags in system settings.
- [ ] Begin creating a new Money Request and proceed to the Confirmation Screen.
- [ ] Tap the dropdown button. Verify neither Categories nor Tags display a `Required` tag.

#### Case 2: Both Categories and Tags Required
- [ ] Enable both categories and tags in system settings.
- [ ] Begin a new Money Request and proceed to the Confirmation Screen.
- [ ] Tap the dropdown button. Verify both Categories and Tags display a `Required` tag.

#### Case 3: Only Categories Required
- [ ] Enable categories and disable tags in system settings.
- [ ] Begin a new Money Request and proceed to the Confirmation Screen.
- [ ] Tap the dropdown button. Verify Categories display a `Required` tag and Tags do not.

#### Case 4: Only Tags Required
- [ ] Disable categories and enable tags in system settings.
- [ ] Begin a new Money Request and proceed to the Confirmation Screen.
- [ ] Tap the dropdown button. Verify Tags display a `Required` tag and Categories do not.

#### Additional Steps for Forced Requirement Display
- This is for development purposes to force display requirements regardless of actual settings.
- [ ] In `MoneyRequestConfirmationList`, set `shouldShowTags` and `shouldShowCategories` to true.
- [ ] In the same file, in the respective MenuItems, set `Boolean(props.policy.requiresCategories)` and `Boolean(props.policy.requiresTags)` to true. Note: This will cause an error if not implemented correctly due to normal guards being disabled.
- Follow the general test steps for each case.

### Offline tests

Same as Tests

### QA Steps

QA should be skipped for now: see [here](https://github.com/Expensify/App/pull/31380#discussion_r1399464659) 

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message: [Slack Message](https://expensify.slack.com/archives/C01GTK53T8Q/p1700064505893339)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>
Android_Required

</details>

<details>
<summary>Android: mWeb Chrome</summary>
mWeb_Android_Required

</details>

<details>
<summary>iOS: Native</summary>
ios_native

</details>

<details>
<summary>iOS: mWeb Safari</summary>
mWeb_iOS_Required

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>
web_Required

</details>

<details>
<summary>MacOS: Desktop</summary>

desktop_Required

</details>

### CODE REVIEW

**Comment by cdanwards** [comment]
There might actually be a change of the word used in the label. Conversation happening in Slack: https://expensify.slack.com/archives/C01GTK53T8Q/p1700064505893339

File: src/components/MoneyRequestConfirmationList.js [hunk 1]
```diff
@@ -699,6 +698,7 @@ function MoneyRequestConfirmationList(props) {
                             titleStyle={styles.flex1}
                             disabled={didConfirm}
                             interactive={!props.isReadOnly}
+                            rightLabel={Permissions.canUseViolations() && Boolean(props.policy.requiresCategory) ? translate('common.required') : ''}
```
Comment (cead22) [inline src/components/MoneyRequestConfirmationList.js:701]: Everything looks good, but I need to make sure we're setting `requiresCategory` and `requiresTag` in the policy object in Onyx

Comment (cdanwards) [reply src/components/MoneyRequestConfirmationList.js:701]: @cead22 is that something to be addressed here or in a separate issue?

Comment (cead22) [reply src/components/MoneyRequestConfirmationList.js:701]: I confirmed this is how this is supposed to be 👍

**Comment by cead22** [comment]
@cdanwards can you update your tests and QA sections
- Make it clear in the test section that all cases should be tested, ie, when neither category or tag is required, when both are required, and when only one is required
- Add steps to configure the policy for the cases above. These settings can be configured in old dot, by going to 
    - For categories: Settings > Workspaces > [the workspace] > Categories > Switch on the toggle "People must categorize expenses", and make sure there's at least 1 enabled category
    - For tags: Settings > Workspaces > [the workspace] > Tags > Switch on the toggle "People must tag expenses", and make sure there's at least 1 enabled tag ("Do you want to use multiple levels of" should be off
- QA steps need to be doable by Applause, and they can't really set anything to return true -- they're running whatever code is on staging

**Comment by cead22** [comment]
@shawnborton can you please take a look and see if we need different styles on the `Required` text?

**Comment by shawnborton** [comment]
Yup, we do need some style adjustments. The text should be using our textSupporting color, and it should be in our label font size (13px).

**Comment by shawnborton** [comment]
Original screenshot for reference:
image

**Comment by situchan** [comment]
@cdanwards after pulling main, app crashes

` ERROR  ReferenceError: Property 'Permissions' doesn't exist`

**Review by situchan — CHANGES_REQUESTED** [review]
use `usePermissions()` hook.
Check #31420 for reference.

File: src/styles/styles.ts [hunk 2]
```diff
@@ -1442,6 +1442,10 @@ const styles = (theme: ThemeColors) =>
             alignItems: 'center',
         },
 
+        rightLabelMenuItem: {
+            width: 65,
```
Comment (situchan) [inline src/styles/styles.ts:1446]: I don't like setting absolute width here. What if text exceeds this width?

Comment (situchan) [reply src/styles/styles.ts:1446]: Just tested. Really bug. No need to set width


Simulator Screenshot - iPhone 15

**Comment by cdanwards** [comment]
@situchan thanks for the review! I'll get it updated.

File: src/components/MenuItem.js [hunk 3]
```diff
@@ -364,6 +365,11 @@ const MenuItem = React.forwardRef((props, ref) => {
                                         />
                                     </View>
                                 )}
+                                {Boolean(props.rightLabel) && (
+                                    <View style={[styles.popoverMenuIcon, styles.rightLabelMenuItem, styles.pointerEventsAuto]}>
```
Comment (situchan) [inline src/components/MenuItem.js:369]:
```suggestion
                                    <View style={[styles.justifyContentCenter, styles.pointerEventsAuto]}>
```

File: src/styles/styles.ts [hunk 4]
```diff
@@ -1442,6 +1442,10 @@ const styles = (theme: ThemeColors) =>
             alignItems: 'center',
         },
 
+        rightLabelMenuItem: {
+            width: 'auto',
+        },
+
```
Comment (situchan) [inline src/styles/styles.ts:1448]: revert this change
```suggestion
```

File: src/components/MenuItem.js (hunk 3 above)
Comment (situchan) [reply src/components/MenuItem.js:369]: Also, what's `styles.pointerEventsAuto` for? I don't see any difference with/without this style.

File: src/components/MenuItem.js [hunk 5]
```diff
@@ -364,6 +365,11 @@ const MenuItem = React.forwardRef((props, ref) => {
...
+                                        <Text>{props.rightLabel}</Text>
```
Comment (situchan) [inline src/components/MenuItem.js:370]: > Yup, we do need some style adjustments. The text should be using our textSupporting color, and it should be in our label font size (13px).

Please add text styles to meet design requirement.

Design:
Screenshot 2023-11-19 at 10 42 26 PM


Current fix:
Screenshot 2023-11-19 at 10 42 40 PM

**Review by situchan — CHANGES_REQUESTED** [review]

File: src/components/MoneyRequestConfirmationList.js (hunk 1 above)
Comment (cdanwards) [reply src/components/MoneyRequestConfirmationList.js]: @cead22 just following up here. I'm rewriting the testing instructions and I've set the proper requirements in old Dot to have categories and tags be required, however I'm not seeing `requiresTags` or `requiresCategories` coming through anywhere on the props from Onyx, in the policy object or otherwise. `props.policyTags.Tag.required` comes through correctly but that's it.

Comment (cead22) [reply src/components/MoneyRequestConfirmationList.js]: Yeah that's fine, maybe we can skip QA steps that depend on those values being true for now, since the backend change for that isn't ready yet

File: src/components/MenuItem.js [hunk 6]
```diff
@@ -364,6 +365,11 @@ const MenuItem = React.forwardRef((props, ref) => {
                                         />
                                     </View>
                                 )}
+                                {Boolean(props.rightLabel) && (
+                                    <View style={styles.justifyContentCenter}>
+                                        <Text style={styles.rightLabelMenuItem}>{props.rightLabel}</Text>
```
Comment (situchan) [inline src/components/MenuItem.js:370]: NAB: `rightLabelMenuItemText` is more correct name

File: src/components/MoneyRequestConfirmationList.js (hunk 1 above)
Comment (cdanwards) [reply src/components/MoneyRequestConfirmationList.js]: Alright @cead22, I updated the test cases and made a note in the QA section that links to this thread.

**Comment by situchan** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

android


</details>

<details>
<summary>Android: mWeb Chrome</summary>

mchrome


</details>

<details>
<summary>iOS: Native</summary>


https://github.com/Expensify/App/assets/108292595/cbae2349-f413-4e23-8149-ce02e64c29d4



</details>

<details>
<summary>iOS: mWeb Safari</summary>


https://github.com/Expensify/App/assets/108292595/0a7c7d82-cb47-476d-9858-5f9ced7f6e78



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>


https://github.com/Expensify/App/assets/108292595/870ab629-817d-4232-92bb-d20dc5502609



</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/108292595/fe3be7df-3cfa-4624-92ff-ec25dd3f8a0d




</details>

**Review by situchan — APPROVED** [review]
:shipit:

**Review by cead22 — APPROVED** [review]

**Comment by izarutskaya** [comment]
Hello @cdanwards, can you please help us, where I can set the `canUseViolations` Permissions to true? Thank you

**Comment by cdanwards** [comment]
@izarutskaya apologies! If you're able to access the `Permissions` file, you can set the function to return true there. I'll update this in the comment above.
